File: apps/ios/Sources/SwabCore/Vault/VaultKeyStore.swift
/// ONB-02: called right after OTP verification, before any classification
/// input is possible. Key never leaves the device (recovery phrase backup is
/// FS-07 OQ-IDT-2, out of onboarding scope). Store id keeps the
/// `swab.vault.key.v1` versioned-name convention shared with the RN
/// reference (`expo-secure-store`) and the Android target (Keystore).
import Foundation

/// Deliberately has no `deleteKey`: destroying the vault key is the VLT-05
/// data-loss event (it orphans every locally-cached ciphertext) and deserves
/// its own reviewed change, not a side effect of the generic `SecureStore`
/// delete added for session-token cleanup (SUG-IOS-012).
public struct VaultKeyStore: Sendable {
    static let storeId = "swab.vault.key.v1"

    private let store: SecureStore

    public init(store: SecureStore) {
        self.store = store
    }

    public func getOrCreateKey() throws -> Data {
        if let existingB64 = try store.get(Self.storeId),
            let existing = Data(base64Encoded: existingB64)
        {
            return existing
        }
        let key = try VaultCrypto.randomKey()
        try store.set(Self.storeId, value: key.base64EncodedString())
        return key
    }
}

File: apps/android/app/src/main/kotlin/com/swab/android/vault/VaultKeyStore.kt
package com.swab.android.vault

import java.security.SecureRandom

/**
 * Vault key storage abstraction (VLT-01, ONB-02). Production is backed by
 * the Android Keystore (see AndroidKeystoreVaultKeyStore in the app's
 * platform layer — never plain SharedPreferences, per android-specialist.md
 * rule 1 and rn-native-handoff.md §2.1). [InMemoryVaultKeyStore] is the JVM
 * unit-test fake so vault/sync logic is testable without a device.
 *
 * Store id keeps the RN reference's versioned-name convention:
 * `swab.vault.key.v1`.
 */
interface VaultKeyStore {
    /**
     * Returns the existing 32-byte vault key, or generates and persists a
     * fresh one. Called right after OTP verification, before any
     * classification input is possible (ONB-02) — the key never leaves the
     * device.
     */
    suspend fun getOrCreateVaultKey(): ByteArray

    companion object {
        const val STORE_ID: String = "swab.vault.key.v1"
        const val KEY_LENGTH_BYTES: Int = 32
    }
}

/** JVM test fake — holds the key in memory only. */
class InMemoryVaultKeyStore : VaultKeyStore {
    private var key: ByteArray? = null

    override suspend fun getOrCreateVaultKey(): ByteArray {
        val existing = key
        if (existing != null) {
            return existing
        }
        val fresh = ByteArray(VaultKeyStore.KEY_LENGTH_BYTES)
        SecureRandom().nextBytes(fresh)
        key = fresh
        return fresh
    }
}
